map_model/src/edits/mod.rs

```diff
@@ -469,33 +454,35 @@ impl Map {
         BTreeSet<TurnID>,
         BTreeSet<IntersectionID>,
     ) {
-        // TODO More efficient ways to do this: given two sets of edits, produce a smaller diff.
-        // Simplest strategy: Remove common prefix.
         let mut effects = EditEffects::new();
 
-        // First undo all existing edits.
-        let mut undo = std::mem::replace(&mut self.edits.commands, Vec::new());
-        undo.reverse();
-        let mut undid = 0;
-        for cmd in &undo {
-            if cmd.undo(&mut effects, self, timer) {
-                undid += 1;
+        // We need to undo() all of the current commands in reverse order, then apply() all of the
+        // new commands. But in many cases, new_edits is just the current edits with a few commands
+        // at the end. So a simple optimization with equivalent behavior is to skip the common
+        // prefix of commands.
+        let mut start_at_idx = 0;
+        for (cmd1, cmd2) in self.edits.commands.iter().zip(new_edits.commands.iter()) {
+            if cmd1 == cmd2 {
+                start_at_idx += 1;
+            } else {
+                break;
             }
         }
-        timer.note(format!("Undid {} / {} existing edits", undid, undo.len()));
+
+        // Undo existing edits
+        for _ in start_at_idx..self.edits.commands.len() {
+            self.edits
+                .commands
+                .pop()
+                .unwrap()
+                .undo()
+                .apply(&mut effects, self, timer);
+        }
 
         // Apply new edits.
-        let mut applied = 0;
-        for cmd in &new_edits.commands {
-            if cmd.apply(&mut effects, self, timer) {
-                applied += 1;
-            }
+        for cmd in &new_edits.commands[start_at_idx..] {
+            cmd.apply(&mut effects, self, timer);
         }
-        timer.note(format!(
```

Comment:
just double checking that you intended to delete the timer note.

Reply:
Yep, it makes STDOUT unnecessarily noisy and doesn't really say anything too interesting